### PR TITLE
storage: remove some derived arbitrary implementations

### PR DESCRIPTION
**Pull request by benesch**
And replace them with hand-rolled implementations. This has solved the
stack overflow we're seeing in storage_command_protobuf_roundtrip in the
past. See #13667.

### Motivation

  * This PR maybe fixes CI flakiness.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

**Comment by frankmcsherry** [comment]
CI failed with what I hope is a transient
```
Error processing tar file(exit status 1): write /tests/mz-repr.test.strconv: no space left on device
```
Retrying.

**Review by frankmcsherry — APPROVED** [review]
I can see what is going on, and it seems not wrong, but don't know enough to know if this is a good or bad fix. If that opinion is already held (e.g. "good!") then LGTM!

File: src/storage/src/types/sources.rs [hunk 1]
```diff
@@ -61,6 +61,30 @@ pub struct IngestionDescription<S = ()> {
...
+            any::<S>(),
+            any::<RelationType>(),
+        )
+            .prop_map(|(desc, source_imports, storage_metadata, typ)| Self {
+                desc,
+                source_imports,
+                storage_metadata,
+                typ,
+            })
+            .boxed()
```
Comment (guswynn) [inline src/storage/src/types/sources.rs:84]: I believe this is the critical change

Comment (benesch) [reply src/storage/src/types/sources.rs:84]: That's what @aalexandrov said, yes!

**Pull request merged**